File: data/code_pairs/cpp/1667.cpp
#ifndef SEGMENT_HPP
#define SEGMENT_HPP

#include <cmath>

namespace segment {
    int whichSegment(double x, double xLower, double xUpper, int iLower, int iUpper) {
        if (x <= xLower) return iLower;
        else if (x >= xUpper) return iUpper;
        else return iLower + static_cast<int>(std::floor(0.5 + (x - xLower) / (xUpper - xLower) * (iUpper - iLower)));
    }
}

#endif // SEGMENT_HPP

File: data/code_pairs/fortran/1667.f90
module segment
  implicit none
contains
  elemental function whichSegment(x, xLower, xUpper, iLower, iUpper) result(i)
    double precision, intent(in) :: x, xLower, xUpper
    integer, intent(in) :: iLower, iUpper
    integer :: i
    i = iLower + floor(0.5D0 + (x - xLower) / (xUpper - xLower) * (iUpper - iLower))
  end function whichSegment
end module segment